packages/ai/core/util/chat-store.ts

```diff
@@ -0,0 +1,602 @@
+import {
+  generateId,
+  generateId as generateIdFunction,
+  parsePartialJson,
+} from '.';
+import type {
+  ReasoningUIPart,
+  TextUIPart,
+  ToolInvocation,
+  ToolInvocationUIPart,
+  UIMessage,
+} from '../types';
+
+interface ChatStoreSubscriber {
+  onChatMessagesChanged?: (id: string) => void;
+  onChatStatusChanged?: (id: string) => void;
+  onChatErrorChanged?: (id: string) => void;
```

Comment:
alternative: `onChatChanged` -- is the granularity needed?

Reply:
can condense! we can pass event type to keep granularity within a single callback